### PR TITLE
Implement support for read-only contract queries in nodeos

### PR DESCRIPTION
**Pull request by jgiszczak**
## Change Description

Implement support for read-only contract queries in nodeos.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [X] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [X] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: plugins/producer_plugin/producer_plugin.cpp [hunk 1]
```diff
@@ -443,13 +444,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+            next(ex);
+
+            if (_trx_trace_failure_log.is_enabled(fc::log_level::debug)) {
+               auto entire_trx = chain_plug->get_entire_trx(trx->get_transaction());
+               fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${entire_trx}", ("entire_trx", entire_trx));
+            }
+         }
+      }
+
+      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, bool read_only, next_function<transaction_trace_ptr> next) {
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:474]: Doesn't look like read_only arg is used.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -443,13 +444,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          schedule_production_loop();
       }
 
-      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, next_function<transaction_trace_ptr> next) {
+      void on_incoming_transaction_sync(const packed_transaction_ptr &trx, bool persist_until_expired, bool read_only, next_function<transaction_trace_ptr> next) {
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:447]: Doesn't look like read_only arg is used.

Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp:447]: Now properly used in constructing transaction metadata.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 1 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp:474]: Now properly used in constructing transaction metadata.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 3]
```diff
@@ -208,6 +206,7 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
       incoming::methods::block_sync::method_type::handle        _incoming_block_sync_provider;
       incoming::methods::blockvault_sync::method_type::handle   _incoming_blockvault_sync_provider;
       incoming::methods::transaction_async::method_type::handle _incoming_transaction_async_provider;
+      incoming::methods::transaction_sync::method_type::handle  _incoming_transaction_sync_provider;
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:209]: Not used

File: plugins/producer_plugin/producer_plugin.cpp [hunk 4]
```diff
@@ -445,28 +444,57 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          schedule_production_loop();
       }
 
-      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, next_function<transaction_trace_ptr> next) {
+      void on_incoming_transaction_sync(const packed_transaction_ptr &trx, bool persist_until_expired, bool read_only, next_function<transaction_trace_ptr> next) {
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:447]: Not used

File: plugins/producer_plugin/producer_plugin.cpp [hunk 5]
```diff
@@ -445,28 +444,57 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
          boost::asio::post(_thread_pool->get_executor(), [self = this, future{std::move(future)}, persist_until_expired,
                                                           next{std::move(next)}, trx]() mutable {
             if( future.valid() ) {
                future.wait();
                app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}, trx{std::move(trx)}]() mutable {
                   auto exception_handler = [self, &next, trx{std::move(trx)}](fc::exception_ptr ex) {
-                     fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
-                            ("txid", trx->id())("a",trx->get_transaction().first_authorizer())("why",ex->what()));
+                    fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
+                            ("txid", trx->id())("why",ex->what()));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:491]: Revert this change. We want `auth`.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 6]
```diff
@@ -491,65 +519,61 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+               if (!trx->read_only) {
+                  _transaction_ack_channel.publish(priority::low, std::pair<fc::exception_ptr, transaction_metadata_ptr>(std::get<fc::exception_ptr>(response), trx));
+               }
                if (_pending_block_mode == pending_block_mode::producing) {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid} : ${why} ",
                         ("block_num", chain.head_block_num() + 1)
                         ("prod", get_pending_block_producer())
                         ("txid", trx->id())
-                        ("a", trx->packed_trx()->get_transaction().first_authorizer())
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:500]: Revert this change, we want `auth`

File: plugins/producer_plugin/producer_plugin.cpp [hunk 7]
```diff
@@ -491,65 +519,61 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                       fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${entire_trace} ",
+                               ("block_num", chain.head_block_num() + 1)
+                               ("prod", get_pending_block_producer())
+                               ("entire_trace", entire_trace));
                   }
                } else {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
                           ("txid", trx->id())
-                          ("a", trx->packed_trx()->get_transaction().first_authorizer())
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:513]: Revert this change, we want `auth`

File: plugins/producer_plugin/producer_plugin.cpp [hunk 8]
```diff
@@ -491,65 +519,61 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                  if (_trx_trace_success_log.is_enabled(fc::log_level::debug)) {
+                       auto entire_trace = chain_plug->get_entire_trx_trace(std::get<transaction_trace_ptr>(response) );
+                       fc_dlog(_trx_trace_success_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is ACCEPTING tx: ${entire_trace}",
+                               ("block_num", chain.head_block_num() + 1)
+                               ("prod", get_pending_block_producer())
+                               ("entire_trace", entire_trace));
+                  }
+               } else {
+                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is ACCEPTING tx: ${txid}",
+                          ("txid", trx->id()));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:570]: We want `auth` here

File: plugins/producer_plugin/producer_plugin.cpp [hunk 9]
```diff
@@ -586,17 +610,12 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
             }
 
-            bool disable_subjective_billing = ( _pending_block_mode == pending_block_mode::producing )
-                                              || ( persist_until_expired && _disable_subjective_api_billing )
-                                              || ( !persist_until_expired && _disable_subjective_p2p_billing );
-
             auto first_auth = trx->packed_trx()->get_transaction().first_authorizer();
             uint32_t sub_bill = 0;
-            if( !disable_subjective_billing )
+            if( _pending_block_mode != pending_block_mode::producing)
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:615]: Revert this change.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 10]
```diff
@@ -586,17 +610,12 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
                sub_bill = _subjective_billing.get_subjective_bill( first_auth, fc::time_point::now() );
 
             auto trace = chain.push_transaction( trx, deadline, trx->billed_cpu_time_us, false, sub_bill );
-            fc_dlog( _trx_failed_trace_log, "Subjective bill for ${a}: ${b} elapsed ${t}us", ("a",first_auth)("b",sub_bill)("t",trace->elapsed));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:599]: Bad merge

File: plugins/producer_plugin/producer_plugin.cpp [hunk 11]
```diff
@@ -758,14 +777,8 @@ void producer_plugin::set_program_options(
...
-         ("disable-subjective-billing", bpo::value<bool>()->default_value(true),
-          "Disable subjective CPU billing for API/P2P transactions")
-         ("disable-subjective-account-billing", boost::program_options::value<vector<string>>()->composing()->multitoken(),
-          "Account which is excluded from subjective CPU billing")
-         ("disable-subjective-p2p-billing", bpo::value<bool>()->default_value(true),
-          "Disable subjective CPU billing for P2P transactions")
-         ("disable-subjective-api-billing", bpo::value<bool>()->default_value(true),
-          "Disable subjective CPU billing for API transactions")
+         ("disable-subjective-billing", bpo::bool_switch()->default_value(false),
+          "Disable subjective billing.")
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:781]: Bad merge

File: plugins/producer_plugin/producer_plugin.cpp [hunk 12]
```diff
@@ -901,25 +914,7 @@ void producer_plugin::plugin_initialize(const boost::program_options::variables_
...
-   if( disable_subjective_billing ) {
-       my->_subjective_billing.disable();
-       ilog( "Subjective CPU billing disabled" );
-   } else if( !my->_disable_subjective_p2p_billing && !my->_disable_subjective_api_billing ) {
-       ilog( "Subjective CPU billing enabled" );
-   } else {
-       if( my->_disable_subjective_p2p_billing ) ilog( "Subjective CPU billing of P2P trxs disabled " );
-       if( my->_disable_subjective_api_billing ) ilog( "Subjective CPU billing of API trxs disabled " );
-   }
+   if( options.at("disable-subjective-billing").as<bool>() ) my->_subjective_billing.disable();
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:917]: Bad merge

File: plugins/producer_plugin/producer_plugin.cpp [hunk 13]
```diff
@@ -988,13 +988,6 @@ void producer_plugin::plugin_initialize(const boost::program_options::variables_
       chain.set_greylist_limit( greylist_limit );
    }
 
-   if( options.count("disable-subjective-account-billing") ) {
-      std::vector<std::string> accounts = options["disable-subjective-account-billing"].as<std::vector<std::string>>();
-      for( const auto& a : accounts ) {
-         my->_subjective_billing.disable_account( account_name(a) );
-      }
-   }
-
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:997]: Revert

File: plugins/producer_plugin/producer_plugin.cpp [hunk 14]
```diff
@@ -1814,15 +1807,17 @@ bool producer_plugin_impl::remove_expired_blacklisted_trxs( const fc::time_point
 }
 
 namespace {
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:1809]: Revert all this to end of file.

File: unittests/contracts/CMakeLists.txt [hunk 15]
```diff
@@ -9,3 +9,8 @@ file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/eosio.wrap/ DESTINATION ${CMAKE_CURRENT_BI
 
 file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/old_versions/v1.6.0-rc3/eosio.bios/ DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/old_versions/v1.6.0-rc3/eosio.bios/)
 file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/old_versions/v1.7.0-develop-preactivate_feature/eosio.bios/ DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/old_versions/v1.7.0-develop-preactivate_feature/eosio.bios/)
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/kv_bios.abi DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/)
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/kv_bios.wasm DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/)
```
Comment (heifner) [inline unittests/contracts/CMakeLists.txt:14]: Is this kv_bios stuff suppose to be part of this PR?

**Review by heifner — CHANGES_REQUESTED** [review]

File: unittests/contracts/CMakeLists.txt (hunk 15 above)
Comment (jgiszczak) [reply unittests/contracts/CMakeLists.txt:14]: Yes, the integration test uses kv tables in the test.  This contract is loaded by the script.

File: plugins/chain_api_plugin/chain_api_plugin.cpp [hunk 16]
```diff
@@ -129,6 +129,7 @@ void chain_api_plugin::plugin_startup() {
       CHAIN_RO_CALL(abi_bin_to_json, 200, http_params_types::params_required),
       CHAIN_RO_CALL(get_required_keys, 200, http_params_types::params_required),
       CHAIN_RO_CALL(get_transaction_id, 200, http_params_types::params_required),
+      CHAIN_RO_CALL_ASYNC(get_contract_query, chain_apis::read_only::get_contract_query_results, 200, http_params_types::params_required),
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_api_plugin/chain_api_plugin.cpp:132]: The name matches the cleos command, which is `cleos get contract_query`.  Should I maintain the match or change it?

File: plugins/chain_interface/include/eosio/chain/plugin_interface.hpp [hunk 17]
```diff
@@ -43,10 +43,11 @@ namespace eosio { namespace chain { namespace plugin_interface {
       }
 
       namespace methods {
-         // synchronously push a block/trx to a single provider
+         // push a block/trx to a single provider
          using block_sync            = method_decl<chain_plugin_interface, bool(const signed_block_ptr&, const std::optional<block_id_type>&), first_provider_policy>;
          using blockvault_sync       = method_decl<chain_plugin_interface, bool(const signed_block_ptr&, bool), first_provider_policy>;
-         using transaction_async     = method_decl<chain_plugin_interface, void(const packed_transaction_ptr&, bool, next_function<transaction_trace_ptr>), first_provider_policy>;
+         using transaction_async     = method_decl<chain_plugin_interface, void(const packed_transaction_ptr&, bool, bool, next_function<transaction_trace_ptr>), first_provider_policy>;
+         using transaction_sync      = method_decl<chain_plugin_interface, void(const packed_transaction_ptr&, bool, bool, next_function<transaction_trace_ptr>), first_provider_policy>;
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_interface/include/eosio/chain/plugin_interface.hpp:50]: Done.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 18]
```diff
@@ -179,6 +179,7 @@ class chain_plugin_impl {
    ,incoming_block_sync_method(app().get_method<incoming::methods::block_sync>())
    ,incoming_blockvault_sync_method(app().get_method<incoming::methods::blockvault_sync>())
    ,incoming_transaction_async_method(app().get_method<incoming::methods::transaction_async>())
+   ,incoming_transaction_sync_method(app().get_method<incoming::methods::transaction_sync>())
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:182]: Done.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 19]
```diff
@@ -210,6 +211,7 @@ class chain_plugin_impl {
    incoming::methods::block_sync::method_type&        incoming_block_sync_method;
    incoming::methods::blockvault_sync::method_type&        incoming_blockvault_sync_method;
    incoming::methods::transaction_async::method_type& incoming_transaction_async_method;
+   incoming::methods::transaction_sync::method_type& incoming_transaction_sync_method;
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:214]: Done.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 20]
```diff
@@ -3183,6 +3185,74 @@ read_only::get_transaction_id_result read_only::get_transaction_id( const read_o
...
+               }
+               const auto& accnt_metadata_obj = db.db().get<account_metadata_object,by_name>( params.account_name );
+               const auto& receipts = db.get_read_mode() == db_read_mode::SPECULATIVE ? db.get_pending_trx_receipts() : deque<transaction_receipt>();
+               vector<transaction_id_type>  pending_transactions;
+               for( transaction_receipt const& receipt : receipts ) {
+                  if( std::holds_alternative<transaction_id_type>(receipt.trx) ) {
+                     pending_transactions.push_back(std::get<transaction_id_type>(receipt.trx));
+                  }
+                  else {
+                     pending_transactions.push_back(std::get<packed_transaction>(receipt.trx).packed_digest());
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:3241]: Done.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 3 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp:209]: Done.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 4 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp:447]: Done.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 21]
```diff
@@ -3183,6 +3185,74 @@ read_only::get_transaction_id_result read_only::get_transaction_id( const read_o
...
+            try {
+               fc::variant output;
+               try {
+                  output = db.to_variant_with_abi( *trx_trace_ptr, abi_serializer::create_yield_function( abi_serializer_max_time ) );
+               } catch( chain::abi_exception& ) {
+                  output = *trx_trace_ptr;
+               }
+               const auto& accnt_metadata_obj = db.db().get<account_metadata_object,by_name>( params.account_name );
+               const auto& receipts = db.get_read_mode() == db_read_mode::SPECULATIVE ? db.get_pending_trx_receipts() : deque<transaction_receipt>();
+               vector<transaction_id_type>  pending_transactions;
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:3235]: Done.

File: plugins/chain_api_plugin/chain_api_plugin.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/chain_api_plugin/chain_api_plugin.cpp:132]: I would argue the cleos command is not indicative of the feature.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 5 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 6 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 7 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 8 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 9 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 10 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 11 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 12 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 13 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 14 above)
Comment (jgiszczak) [reply plugins/producer_plugin/producer_plugin.cpp]: Bad merge repaired.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 22]
```diff
@@ -3183,6 +3185,74 @@ read_only::get_transaction_id_result read_only::get_transaction_id( const read_o
...
+
+            try {
+               fc::variant output;
+               try {
+                  output = db.to_variant_with_abi( *trx_trace_ptr, abi_serializer::create_yield_function( abi_serializer_max_time ) );
+               } catch( chain::abi_exception& ) {
+                  output = *trx_trace_ptr;
+               }
+               const auto& accnt_metadata_obj = db.db().get<account_metadata_object,by_name>( params.account_name );
+               const auto& receipts = db.get_read_mode() == db_read_mode::SPECULATIVE ? db.get_pending_trx_receipts() : deque<transaction_receipt>();
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:3232]: Fixed.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 23]
```diff
@@ -3183,6 +3185,74 @@ read_only::get_transaction_id_result read_only::get_transaction_id( const read_o
...
+      } EOS_RETHROW_EXCEPTIONS(chain::packed_transaction_type_exception, "Invalid packed transaction")
+
+      auto trx_trace = fc_create_trace_with_id("TransactionReadOnly", input_trx->id());
+      auto trx_span = fc_create_span(trx_trace, "HTTP Received");
+      fc_add_tag(trx_span, "trx_id", input_trx->id());
+      fc_add_tag(trx_span, "method", "send_transaction");
+
+      app().get_method<incoming::methods::transaction_async>()(input_trx, true, true,
+            [this, token=trx_trace.get_token(), input_trx, params, next]
+            (const std::variant<fc::exception_ptr, transaction_trace_ptr>& result) -> void {
```
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_plugin/chain_plugin.cpp:3203]: Added support for returning failure traces.

File: plugins/chain_api_plugin/chain_api_plugin.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (jgiszczak) [reply plugins/chain_api_plugin/chain_api_plugin.cpp]: Changed API to `send_ro_transaction`.

File: unittests/contracts/read_only_query_tests.abi [hunk 24]
```diff
@@ -0,0 +1,105 @@
+{
```
Comment (victorj8) [inline unittests/contracts/read_only_query_tests.abi:1]: I think it would also be a good idea to add the smart contract(cpp file) used to generate this abi and wasm.

Comment (jgiszczak) [reply unittests/contracts/read_only_query_tests.abi:1]: The source file for the test contract is part of `eosio.cdt`.  We've decided to avoid versioning problems likely to arise from keeping two copies of it.

Comment (victorj8) [reply unittests/contracts/read_only_query_tests.abi:1]: got it, i think the eosio.cdt is https://github.com/EOSIO/eosio.cdt/pull/1068, let me know if it is not

**Review by larryk85 — APPROVED** [review]

File: plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp [hunk 25]
```diff
@@ -565,6 +565,24 @@ class read_only {
       };
    }
 
+   struct push_ro_transaction_params {
+      name account_name;
+      bool return_failure_traces;
```
Comment (heifner) [inline plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp:570]: Should initialize to either `true` or `false`.

**Review by heifner — APPROVED** [review]
